Svc/SeqDispatcher/SeqDispatcher.hpp

```diff
@@ -0,0 +1,95 @@
+// ======================================================================
+// \title  SeqDispatcher.hpp
+// \author zimri.leisher
+// \brief  hpp file for SeqDispatcher component implementation class
+// ======================================================================
+
+#ifndef SeqDispatcher_HPP
+#define SeqDispatcher_HPP
+
+#include "Svc/SeqDispatcher/SeqDispatcherComponentAc.hpp"
+#include "Svc/SeqDispatcher/SeqDispatcher_CmdSequencerStateEnumAc.hpp"
+#include "FppConstantsAc.hpp"
+#include "Fw/Types/WaitEnumAc.hpp"
+#include "Fw/Types/StringBase.hpp"
+
+namespace Svc {
+
+class SeqDispatcher : public SeqDispatcherComponentBase {
+  public:
+    // ----------------------------------------------------------------------
+    // Construction, initialization, and destruction
+    // ----------------------------------------------------------------------
+
+    //! Construct object SeqDispatcher
+    //!
+    SeqDispatcher(const char* const compName /*!< The component name*/
+    );
+
+    //! Destroy object SeqDispatcher
+    //!
+    ~SeqDispatcher();
+
+  PROTECTED:
+
+    //! Handler for input port seqDoneIn
+    void
+    seqDoneIn_handler(NATIVE_INT_TYPE portNum,         //!< The port number
+                      FwOpcodeType opCode,             //!< Command Op Code
+                      U32 cmdSeq,                      //!< Command Sequence
+                      const Fw::CmdResponse& response  //!< The command response argument
+    );
+
+    //! Handler for input port seqStartIn
+    void seqStartIn_handler(NATIVE_INT_TYPE portNum, //!< The port number
+                            const Fw::StringBase& fileName //!< The sequence file
+    );
+
+    //! Handler for input port seqRunIn
+    void seqRunIn_handler(NATIVE_INT_TYPE portNum, //!< The port number
+                          const Fw::StringBase& fileName //!< The sequence file
+    );
+
+  PRIVATE:
+
+    // number of sequences dispatched (successful or otherwise)
+    U32 m_dispatchedCount = 0;
+    // number of errors from dispatched sequences (CmdResponse::EXECUTION_ERROR)
+    U32 m_errorCount = 0;
+    // number of sequencers in state AVAILABLE
+    U32 m_sequencersAvailable = SeqDispatcherSequencerPorts;
+
+    struct DispatchEntry {
+        FwOpcodeType opCode;  //!< opcode of entry
+        U32 cmdSeq;
+        // store the state of each sequencer
+        SeqDispatcher_CmdSequencerState state;
```

Comment:
Next month (September 2024) Fprime and FPP will be released to directly support state machines within a component.
The Phase 1 delivery will allow developers to instantiate an externally defined state machine in FPP.  The externally defined state machine can be specified in the open source QM modeling tool or in PlantUML text and the STARS (https://github.com/JPLOpenSource/STARS) state machine autocoder will generate the code.

My suggestion is to specify the state of the Sequencer as a state machine model that is then integrated into the component.  Here is a cut at the state machine specified using the QM modeling tool.  The developer would then write the code snippets to implement the state machine functions.
![image](https://github.com/user-attachments/assets/c6602d25-59c3-4569-aa62-3d52fbe82c0e)

In addition the state machine can just be specified in PlantUML and then run through the same STARS autocoder. 

```
@startuml

[*] --> Available

state Available {
    Available:Internal: Done / UnknownSeqFinished(); incrSeqAvailable(e) 
}

state Running {
    state c1 <<choice>>
    [*] --> NonBlock
    Running:Internal: Start [seqRunningNotFile(e)] / unexpectedSeqStart(e)
    Running:Internal: Run / invalidSequencer(e); sendExecutionError(e)

    state NonBlock {
    }

    state Block {
    }
}

Available --> c1 : Run /seqRunOut(e); decSeqAvailable(e)
c1 --> Block
c1 --> NonBlock : [NoWait(e)]
NonBlock --> Available : Done / incrSeqAvailable(e)
Block --> Available : Done / sendCmdResponse(e); incrSeqAvailable(e)
Available --> Running : Start / decSeqAvailable(e)



@enduml

```

The graphics of the PlantUML can be automatically rendered by python -plantuml or in Visual Studio as:

![image](https://github.com/user-attachments/assets/f88da2d4-a7d2-4b52-a756-6e58bd25e982)

In Phase 2 of the state machine delivery, the state machine model will be specified directly in the component FPP language eliminating the need to use STARS to define the external model or generate the code.

If you do not want to go down this route then at a minimum include the state machine diagrams in the SDD.

Reply:
Since this component is in its final review (and needed by customers).  I recommend we track this as a issue to work in the future.

Reply:
I recommend that the model diagram be placed in the SDD.

Reply:
I will add the model diagrams you produced to the SDD, if that's okay with you.